### PR TITLE
object inspection: produce deterministic descriptions for nested collection datastructures

### LINKED ISSUE #11198: util.inspect.object_description: does not emit reliable ordering for a set nested within another collection
### Summary
Differences appear in some `sphinx` v5.3.0 generated `set` object descriptions for `alembic` v1.8.1, as demonstrated by [recent results visible on the Reproducible Builds diffoscope dashboard](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/diffoscope-results/alembic.html).

Arguably it could make sense for code authors to intentionally write `set` elements in their code files in a way that does not correspond to their computed sort order -- as a means to communicate with human readers about abstract ideas that aren't relevant to computers at runtime, for example.

However, the current behaviour does result in non-reproducible documentation output.

### Details
In particular, the ordering of a class attribute with a value that contains a set-within-a-tuple seems unreliable across differing builds:

https://github.com/sqlalchemy/alembic/blob/a968c9d2832173ee7d5dde50c7573f7b99424c38/alembic/ddl/impl.py#L90

... is emitted variously as ...

```
<span·class="pre">({'NUMERIC',</span>·<span·class="pre">'DECIMAL'},)</span>
```

... or ...

```
<span·class="pre">({'DECIMAL',</span>·<span·class="pre">'NUMERIC'},)</span>
```

cc @lamby who has been [investigating a fix on the reproducible-builds mailing list](https://lists.reproducible-builds.org/pipermail/rb-general/2023-February/002862.html).

### How to Reproduce

It is not yet clear to me exactly what circumstances cause the ordering of elements to vary - and it's OK not to proceed until that's figured out (maybe not a blocker, but it would be nice to have confidence about the cause).

From searching around on previous issues while writing up this bugreport: I wonder if this could be an edge-case for / follow-up to #4834.

### Environment Information

Although these build log links are somewhat ephemeral, the system environment details for two builds that produce differing output are visible at:

- https://tests.reproducible-builds.org/debian/rbuild/unstable/amd64/alembic_1.8.1-2.rbuild.log.gz
- https://tests.reproducible-builds.org/debian/logs/unstable/amd64/alembic_1.8.1-2.build2.log.gz

### Sphinx extensions

```python
https://github.com/sqlalchemy/alembic/blob/rel_1_8_1/docs/build/conf.py#L36-L42


sphinx.ext.autodoc
sphinx.ext.intersphinx
changelog
sphinx_paramlinks
sphinx_copybutton
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix

### Purpose
- Improve deterministic reproducibility of documentation

### Detail
- `util.inspect.object_description` already attempts to sort collections, but this can fail
- This changeset handles the failure case by using string-based object descriptions as a fallback deterministic sort ordering
- This changeset also adds recursive object-description calls for `list` and `tuple` datatypes

### Relates
- May resolve #11198

### CODE REVIEW

**Comment by jayaddison** [comment]
(force-pushed to edit my commit authorship because I remembered that puzzling over this originated from use of `alembic` at work)

**Review by AA-Turner — APPROVED** [review]
Is this ready to merge?

A

**Comment by jayaddison** [comment]
@AA-Turner I think so, but I'd like to wait until the other changes I have pending for 6.2 are sorted out before merging this one.

(it takes me quite a while to regain context when re-reading about each change, and so I think I can better-support a small number of changes introduced gradually)

**Comment by jayaddison** [comment]
I'll try to be a bit more precise than my previous message: I'd like to wait until _after_ the 6.2.0 release, before this can be merged.

**Comment by jayaddison** [comment]
@AA-Turner do we have any mechanisms to monitor memory usage and/or runtime performance impact of pull requests?

I'm now comfortable with this being merged, but would be more confident if we had some performance statistics available (in particular, it's the collection of `seen` object identifiers in one of the post-approval commits that I'm wondering about).

**Comment by david-a-wheeler** [comment]
I'm really glad to see this work to deterministically reproduce documents. Thank you!

**Comment by jayaddison** [comment]
> I'm now comfortable with this being merged, but would be more confident if we had some performance statistics available (in particular, it's the collection of `seen` object identifiers in one of the post-approval commits that I'm wondering about).

Maybe slightly off-topic, but from looking into `pytest`-compatible performance testing tools: [`pytest-austin`](https://github.com/P403n1x87/pytest-austin/) looks potentially promising to me.  I might try it out for some unrelated projects, although would also be glad to help try it out here if that'd be useful.

The `pytest-austin` repository itself hasn't had much commit/contributor activity, but I don't think that's a negative indicator; it pulls-in a related dependency, [`austin`](https://github.com/P403n1x87/austin/), to perform the measurement tasks, and that does appear to be actively maintained.

**Comment by paravoid** [comment]
> I'll try to be a bit more precise than my previous message: I'd like to wait until _after_ the 6.2.0 release, before this can be merged.

Should this be merged now then? Thanks!

**Comment by jayaddison** [comment]
This can be merged in my opinion - I've re-reviewed the changes to confirm my sense about that.  Whether it should be merged remains up to the maintainers.

File: sphinx/util/inspect.py [hunk 1]
```diff
@@ -357,43 +357,59 @@ def safe_getattr(obj: Any, name: str, *defargs: Any) -> Any:
...
 
 
-def object_description(object: Any) -> str:
+def object_description(object: Any, seen: frozenset = None) -> str:
     """A repr() implementation that returns text safe to use in reST context."""
+    seen = seen or frozenset()
     if isinstance(object, dict):
+        if id(object) in seen:
+            return "dict(...)"
+        seen |= frozenset([id(object)])
```
Comment (AA-Turner) [inline sphinx/util/inspect.py]:
```suggestion
        seen |= {id(object)}
```

File: sphinx/util/inspect.py [hunk 2]
```diff
@@ -357,43 +357,59 @@ def safe_getattr(obj: Any, name: str, *defargs: Any) -> Any:
...
         try:
             sorted_keys = sorted(object)
         except Exception:
```
Comment (AA-Turner) [inline sphinx/util/inspect.py]:
```suggestion
        except TypeError:
```

File: sphinx/util/inspect.py [hunk 3]
```diff
@@ -357,43 +357,59 @@ def safe_getattr(obj: Any, name: str, *defargs: Any) -> Any:
...
             items = ("%s: %s" %
-                     (object_description(key), object_description(object[key]))
+                     (object_description(key, seen), object_description(object[key], seen))
                      for key in sorted_keys)
             return "{%s}" % ", ".join(items)
     elif isinstance(object, set):
-        set_descr = (object_description(x) for x in object)
+        if id(object) in seen:
+            return "set(...)"
+        seen |= frozenset([id(object)])
```
Comment (AA-Turner) [inline sphinx/util/inspect.py]:
```suggestion
        seen |= {id(object)}
```

File: sphinx/util/inspect.py [hunk 4]
```diff
@@ -357,43 +357,59 @@ def safe_getattr(obj: Any, name: str, *defargs: Any) -> Any:
...
-        else:
-            return "{%s}" % ", ".join(sorted_set_descr)
+            # Cannot sort set values, fall back to using descriptions as a sort key
+            sorted_values = sorted(object, key=lambda x: object_description(x, seen))
+        return "{%s}" % ", ".join(object_description(x, seen) for x in sorted_values)
     elif isinstance(object, frozenset):
-        frozenset_descr = (object_description(x) for x in object)
+        if id(object) in seen:
+            return "frozenset(...)"
+        seen |= frozenset([id(object)])
```
Comment (AA-Turner) [inline sphinx/util/inspect.py]:
```suggestion
        seen |= {id(object)}
```

File: sphinx/util/inspect.py [hunk 5]
```diff
@@ -357,43 +357,59 @@ def safe_getattr(obj: Any, name: str, *defargs: Any) -> Any:
...
-            ", ".join(object_description(x) for x in object),
-            "," if len(object) == 1 else ""
+            ", ".join(object_description(x, seen) for x in object),
+            "," if len(object) == 1 else "",
         )
     elif isinstance(object, list):
-        return "[%s]" % ", ".join(object_description(x) for x in object)
+        if id(object) in seen:
+            return "list(...)"
+        seen |= frozenset([id(object)])
```
Comment (AA-Turner) [inline sphinx/util/inspect.py]:
```suggestion
        seen |= {id(object)}
```

File: sphinx/util/inspect.py [hunk 6]
```diff
@@ -357,35 +357,61 @@ def safe_getattr(obj: Any, name: str, *defargs: Any) -> Any:
         raise AttributeError(name) from exc
 
 
-def object_description(object: Any) -> str:
-    """A repr() implementation that returns text safe to use in reST context."""
+def object_description(object: Any, seen: frozenset = frozenset()) -> str:
+    """A repr() implementation that returns text safe to use in reST context.
+
+    Maintains a set of 'seen' objectIDs to detect and avoid infinite recursion.
```
Comment (AA-Turner) [inline sphinx/util/inspect.py]:
```suggestion
    Maintains a set of 'seen' object IDs to detect and avoid infinite recursion.
```

File: tests/test_util_inspect.py [hunk 7]
```diff
@@ -503,12 +503,35 @@ def test_set_sorting():
...
+
+    set_ = set(MyEnum)
+    description = inspect.object_description(set_)
+    assert description == "{MyEnum.a, MyEnum.b, MyEnum.c}"
+
+
 def test_set_sorting_fallback():
     set_ = {None, 1}
     description = inspect.object_description(set_)
     assert description in ("{1, None}", "{None, 1}")
```
Comment (AA-Turner) [inline tests/test_util_inspect.py]: Why can't we guarantee one of these two?

File: tests/test_util_inspect.py [hunk 8]
```diff
@@ -521,6 +544,38 @@ def test_frozenset_sorting_fallback():
     assert description in ("frozenset({1, None})", "frozenset({None, 1})")
```
Comment (AA-Turner) [inline tests/test_util_inspect.py]: Why can't we guarantee one of these two?

**Comment by david-a-wheeler** [comment]
Thanks so much for considering this! Making descriptions deterministic will make it easier to determine if a given package really *was* generated by its claimed source. I hope this (or something like it) will be merged.

**Comment by jayaddison** [comment]
Thanks @david-a-wheeler - to be transparent: I can't guarantee that this resolves the only remaining source of non-replicable documentation; there may be other scenarios where generated documentation can differ.  Similarly, it may not be the best possible approach to making these descriptions deterministic; but it's an improvement, I think.

**Comment by david-a-wheeler** [comment]
@jayaddison - I understand. Improvements are improvements, though!

**Comment by AA-Turner** [comment]
Thanks @jayaddison @lamby @paravoid!

A

**Comment by jayaddison** [comment]
Thank you, @AA-Turner!

**Pull request merged**